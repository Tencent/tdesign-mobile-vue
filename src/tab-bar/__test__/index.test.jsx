import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import TabBar from '../tab-bar';
import TabBarItem from '../tab-bar-item';
import { ref } from 'vue';

const list = [
  {
    name: 'label_1',
    text: '标签一',
    icon: undefined,
    value: '1',
  },
  {
    name: 'label_2',
    text: '标签二',
    icon: undefined,
    value: '2',
  },
  {
    name: 'label_3',
    text: '此处展开',
    icon: undefined,
    value: '3',
    subTabBar: [
      {
        name: 'spread_1',
        text: '展开项一',
        icon: undefined,
        value: '3-1',
      },
      {
        name: 'spread_2',
        text: '展开项二',
        icon: undefined,
        value: '3-2',
      },
      {
        name: 'spread_3',
        text: '展开项三',
        icon: undefined,
        value: '3-3',
      },
    ],
  },
];

describe('TabBar', () => {
  describe('props', () => {
    it('bordered', async () => {
      const wrapper = mount(TabBar, {
        shallow: true,
      });
      expect(wrapper.classes()).toContain('t-tab-bar--bordered');

      await wrapper.setProps({ bordered: false });
      expect(wrapper.classes()).not.toContain('t-tab-bar--bordered');
    });

    it('fixed', async () => {
      const wrapper = mount(TabBar, {
        shallow: true,
      });
      expect(wrapper.classes()).toContain('t-tab-bar--fixed');

      await wrapper.setProps({ fixed: false });
      expect(wrapper.classes()).not.toContain('t-tab-bar--fixed');
    });

    it('defaultValue', async () => {
      const wrapper = mount({
        render: () => (
          <TabBar defaultValue={'2'}>
            {list.map((item) => {
              return <TabBarItem {...item}>{item.text}</TabBarItem>;
            })}
          </TabBar>
        ),
      });

      expect(wrapper.findAll('.t-tab-bar-item')).toHaveLength(list.length);
      // expect(
      //   wrapper
      //     .findAllComponents(TabBarItem)
      //     .find((item) => item.vm.isChecked)
      //     .text(),
      // ).toEqual('标签二');
    });

    it('item without value', async () => {
      const wrapper = mount({
        render() {
          return <TabBar value={0}>
            <TabBarItem></TabBarItem>
          </TabBar>
        }
      })

      const item = wrapper.findComponent('.t-tab-bar-item');
      // expect(item.vm.isChecked).toBeTruthy();
    })
  });

  describe('events', () => {
    it('click', async () => {
      const onChange = vi.fn();
      const value = ref('1');
      const wrapper = mount({
        render: () => (
          <TabBar v-model={value.value} onChange={onChange}>
            {list.map((item) => {
              return <TabBarItem {...item} />;
            })}
          </TabBar>
        ),
      });

      // 从 bar1 切到 bar2
      const bar2 = wrapper.find('[name="label_2"] > .t-tab-bar-item__content');
      expect(onChange).toHaveBeenCalledTimes(0);
      await bar2.trigger('click');
      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenLastCalledWith('2');

      // 从 bar2 切到 bar3
      const bar3 = wrapper.find('[name="label_3"] > .t-tab-bar-item__content');
      await bar3.trigger('click');
      expect(onChange).toHaveBeenCalledTimes(2);
      expect(onChange).toHaveBeenLastCalledWith(['3']);

      // 点击 spread
      const bar3_1 = wrapper.find('[name="label_3"] .t-tab-bar-item__spread-item');
      await bar3_1.trigger('click');
      expect(onChange).toHaveBeenCalledTimes(3);
      expect(onChange).toHaveBeenLastCalledWith(['3', '3-1']);
      expect(wrapper.find('[name="label_3"]').find('.t-tab-bar-item__spread-item').exists()).toBe(false);
    });
  });
});
