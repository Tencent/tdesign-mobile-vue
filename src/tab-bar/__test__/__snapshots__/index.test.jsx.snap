// Vitest Snapshot v1

exports[`TabBar > props > snapshot 1`] = `
<div
  class="t-tab-bar t-tab-bar--bordered t-tab-bar--fixed"
  role="tablist"
>
  
  
  <div
    aria-label="TabBar"
    aria-selected="false"
    class="t-tab-bar-item t-no-border"
    name="label_1"
    role="tab"
    text="标签一"
  >
    <div
      class="t-tab-bar-item__content"
    >
      <div
        class="t-tab-bar-item__icon"
      >
        <div
          class="t-badge"
        >
          <div
            class="t-badge__inner t-badge--has-children t-badge--medium t-badge--circle"
            style="right: 0px; top: 5px;"
          >
            1
          </div>
          
          <svg
            class="t-icon t-icon-app"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.25 1.75a3 3 0 100 6 3 3 0 000-6zm-2 3a2 2 0 114 0 2 2 0 01-4 0zM2 3a1 1 0 011-1h3.5a1 1 0 011 1v3.5a1 1 0 01-1 1H3a1 1 0 01-1-1V3zm1 0v3.5h3.5V3H3zM2 9.5a1 1 0 011-1h3.5a1 1 0 011 1V13a1 1 0 01-1 1H3a1 1 0 01-1-1V9.5zm1 0V13h3.5V9.5H3zM8.5 9.5a1 1 0 011-1H13a1 1 0 011 1V13a1 1 0 01-1 1H9.5a1 1 0 01-1-1V9.5zm1 3.5H13V9.5H9.5V13z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
          
        </div>
      </div>
      <!--v-if-->
    </div>
    <transition-stub>
      <!--v-if-->
    </transition-stub>
  </div>
  <div
    aria-label="TabBar"
    aria-selected="false"
    class="t-tab-bar-item t-no-border"
    name="label_2"
    role="tab"
    text="标签二"
  >
    <div
      class="t-tab-bar-item__content"
    >
      <div
        class="t-tab-bar-item__icon"
      >
        <div
          class="t-badge"
        >
          <div
            class="t-badge__inner t-badge--has-children t-badge--medium t-badge--circle"
            style="right: 0px; top: 5px;"
          >
            1
          </div>
          
          <svg
            class="t-icon t-icon-app"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.25 1.75a3 3 0 100 6 3 3 0 000-6zm-2 3a2 2 0 114 0 2 2 0 01-4 0zM2 3a1 1 0 011-1h3.5a1 1 0 011 1v3.5a1 1 0 01-1 1H3a1 1 0 01-1-1V3zm1 0v3.5h3.5V3H3zM2 9.5a1 1 0 011-1h3.5a1 1 0 011 1V13a1 1 0 01-1 1H3a1 1 0 01-1-1V9.5zm1 0V13h3.5V9.5H3zM8.5 9.5a1 1 0 011-1H13a1 1 0 011 1V13a1 1 0 01-1 1H9.5a1 1 0 01-1-1V9.5zm1 3.5H13V9.5H9.5V13z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
          
        </div>
      </div>
      <!--v-if-->
    </div>
    <transition-stub>
      <!--v-if-->
    </transition-stub>
  </div>
  <div
    aria-haspopup="true"
    aria-label="TabBar"
    aria-selected="false"
    class="t-tab-bar-item t-no-border"
    name="label_3"
    role="tab"
    text="此处展开"
  >
    <div
      class="t-tab-bar-item__content"
    >
      <div
        class="t-tab-bar-item__icon"
      >
        <div
          class="t-badge"
        >
          <div
            class="t-badge__inner t-badge--has-children t-badge--medium t-badge--circle"
            style="right: 0px; top: 5px;"
          >
            1
          </div>
          
          <svg
            class="t-icon t-icon-app"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.25 1.75a3 3 0 100 6 3 3 0 000-6zm-2 3a2 2 0 114 0 2 2 0 01-4 0zM2 3a1 1 0 011-1h3.5a1 1 0 011 1v3.5a1 1 0 01-1 1H3a1 1 0 01-1-1V3zm1 0v3.5h3.5V3H3zM2 9.5a1 1 0 011-1h3.5a1 1 0 011 1V13a1 1 0 01-1 1H3a1 1 0 01-1-1V9.5zm1 0V13h3.5V9.5H3zM8.5 9.5a1 1 0 011-1H13a1 1 0 011 1V13a1 1 0 01-1 1H9.5a1 1 0 01-1-1V9.5zm1 3.5H13V9.5H9.5V13z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
          
        </div>
      </div>
      <!--v-if-->
    </div>
    <transition-stub>
      <!--v-if-->
    </transition-stub>
  </div>
  
  
</div>
`;
